tests/python/unittest/test_io.py

```diff
@@ -88,79 +89,122 @@ def test_Cifar10Rec():
+        batch_list.append(batch.data[0].asnumpy())
+    dataiter = mx.io.NDArrayIter(data, labels, 1, True)
+    idx_list = dataiter.idx
+    i = 0
     for batch in dataiter:
-        label = batch.label[0].asnumpy().flatten()
-        assert((batch.data[0].asnumpy()[:, 0, 0] == label).all())
-        for i in range(label.shape[0]):
-            labelcount[int(label[i])] += 1
+        # check if each data point have been shuffled to corresponding positions
+        assert np.array_equal(batch.data[0].asnumpy(), batch_list[idx_list[i]])
+        i += 1
 
-    for i in range(10):
-        if i == 0:
-            assert(labelcount[i] == 124)
-        else:
-            assert(labelcount[i] == 100)
+
+def test_NDArrayIter():
+    dtype_list = ['NDArray', 'ndarray']
+    for dtype in dtype_list:
+        data, labels = _init_NDArrayIter_data(dtype_list)
+        _test_last_batch_handle(data, labels)
+        _test_last_batch_handle([data, data], labels)
+        _test_last_batch_handle(data, [])
+        _test_last_batch_handle(data)
+        _test_shuffle(data, labels)
+        _test_shuffle([data, data], labels)
+        _test_shuffle(data, [])
+        _test_shuffle(data)
 
 
 def test_NDArrayIter_h5py():
     if not h5py:
         return
 
-    data = np.ones([1000, 2, 2])
-    label = np.ones([1000, 1])
-    for i in range(1000):
-        data[i] = i / 100
-        label[i] = i / 100
+    data, labels = _init_NDArrayIter_data('ndarray')
 
     try:
-        os.remove("ndarraytest.h5")
+        os.remove('ndarraytest.h5')
     except OSError:
         pass
-    with h5py.File("ndarraytest.h5") as f:
-        f.create_dataset("data", data=data)
-        f.create_dataset("label", data=label)
-
-        dataiter = mx.io.NDArrayIter(
-            f["data"], f["label"], 128, True, last_batch_handle='pad')
-        batchidx = 0
-        for batch in dataiter:
-            batchidx += 1
-        assert(batchidx == 8)
-
-        dataiter = mx.io.NDArrayIter(
-            f["data"], f["label"], 128, False, last_batch_handle='pad')
-        labelcount = [0 for i in range(10)]
-        for batch in dataiter:
-            label = batch.label[0].asnumpy().flatten()
-            assert((batch.data[0].asnumpy()[:, 0, 0] == label).all())
-            for i in range(label.shape[0]):
-                labelcount[int(label[i])] += 1
-
+    with h5py.File('ndarraytest.h5') as f:
+        f.create_dataset('data', data=data)
+        f.create_dataset('label', data=labels)
+        
+        _test_last_batch_handle(f['data'], f['label'])
+        _test_last_batch_handle(f['data'], [])
+        _test_last_batch_handle(f['data'])
     try:
         os.remove("ndarraytest.h5")
     except OSError:
         pass
 
-    for i in range(10):
-        if i == 0:
-            assert(labelcount[i] == 124)
-        else:
-            assert(labelcount[i] == 100)
+
+def _test_NDArrayIter_csr(csr_iter, csr_iter_empty_list, csr_iter_None, num_rows, batch_size):
```

Comment:
does this cover all failure case last time? Just wanna be super careful  with all existing use cases. We can contact those guys if we need more test cases.

Reply:
sure!  It passes the dcgan example and pixel to pixel but it would be great if we make sure it covers all the use cases. Let do it

Reply:
@stu1130 - You tested and added tests from AirBnB case that was broken earlier?

Reply:
@sandeep-krishnamurthy @zhreshold 
data shape is (5000,1,28,28), the batch size on the test that failed was 64
```
data_iter = mx.io.NDArrayIter(x, y, batch_size=batch_size)
for _ in six.moves.xrange(self.epoch):
    data_iter.reset()
    for batch in data_iter:
         # split the data batch and load them on GPUs
         data = mx.gluon.utils.split_and_load(batch.data[0], self.ctx)
         label = mx.gluon.utils.split_and_load(batch.label[0], self.ctx)
         # compute gradient
         with mx.autograd.record():
             losses = [self.loss(self.net(x_), y_) for x_, y_ in zip(data, label)]
             for l in losses:
                 l.backward()
             # update parameters
         self.trainer.step(self.batch_size)
```

Reply:
Thanks looks good to me.

Reply:
add (5000, 1, 28, 28) data source
https://github.com/apache/incubator-mxnet/blob/4a7c9018b392c5f8bace62ff92032c32e803cf50/tests/python/unittest/test_io.py#L112-L113